client/onboarding/add-business-info-task/index.js

```diff
@@ -0,0 +1,171 @@
+import WizardTaskContext from 'wcpay/additional-methods-setup/wizard/task/context';
+import OnboardingSelectControl from 'wcpay/components/onboarding-select-control';
+import { useBusinessTypes } from 'wcpay/data/onboarding';
+import strings from '../strings';
+
+const AddBusinessInfo = () => {
+	const { setCompleted } = useContext( WizardTaskContext );
+	const { businessTypes, isLoading } = useBusinessTypes();
+	const {
+		connect: { availableCountries, country },
+	} = wcpaySettings;
+
+	const countries = Object.keys( availableCountries ).map(
+		( countryCode ) => {
+			return {
+				label: availableCountries[ countryCode ],
+				value: countryCode,
+			};
+		}
+	);
+
+	const [ businessCountry, setBusinessCountry ] = useState( country );
+	const [ businessType, setBusinessType ] = useState();
+	const [ businessStructure, setBusinessStructure ] = useState();
+	const [ displayStructures, setDisplayStructures ] = useState( false );
+
+	const getBusinessTypesForCountry = ( countryCode ) => {
+		if ( isLoading || ! businessTypes.hasOwnProperty( 'data' ) ) {
+			return false;
+		}
+
+		return businessTypes.data.hasOwnProperty( countryCode )
+			? businessTypes.data[ countryCode ]
+			: false;
+	};
+
+	const formatBusinessTypes = ( countryCode ) => {
+		const businessTypeData = getBusinessTypesForCountry( countryCode );
+
+		if ( isLoading || ! businessTypeData ) {
+			return [];
+		}
+
+		return Object.keys( businessTypeData ).map( ( key ) => ( {
+			name: strings.businessTypes[ key ],
+			description: 'Test',
+			key: key,
+		} ) );
+	};
+
+	const shouldDisplayStructures = ( countryCode, type ) => {
+		const businessTypeData = getBusinessTypesForCountry( countryCode );
+
+		if ( isLoading || ! businessTypeData ) {
+			return false;
+		}
+
+		return businessTypeData.hasOwnProperty( type )
+			? 0 < businessTypeData[ type ].length
+			: false;
+	};
+
+	const formatBusinessStructures = ( countryCode, type ) => {
+		const businessTypeData = getBusinessTypesForCountry( countryCode );
+
+		if ( isLoading || ! businessTypeData ) {
+			return false;
+		}
+
+		return businessTypeData.hasOwnProperty( type )
+			? businessTypeData[ type ]
+			: [];
+	};
+
+	const handleBusinessTypeUpdate = ( type ) => {
+		setBusinessType( type );
+
+		if ( shouldDisplayStructures( businessCountry, type ) ) {
+			formatBusinessStructures( businessCountry, type );
+			setDisplayStructures( true );
+		} else {
+			setDisplayStructures( false );
+			setCompleted( true );
+		}
+	};
+
+	const handleBusinessStructureUpdate = ( structure ) => {
+		setBusinessStructure( structure );
+		setCompleted( true );
+	};
+
+	const handleBusinessCountryUpdate = ( countryCode ) => {
+		setBusinessCountry( countryCode );
+		formatBusinessTypes( countryCode );
+		setCompleted( false );
+	};
+
+	return (
+		<WizardTaskItem
+			className="complete-business-info-task"
+			index={ 1 }
+			title={ strings.onboarding.heading }
+		>
+			{ isLoading ? <p>Loading...</p> : '' }
```

Comment:
You can use the `<LoadableBlock/>` component to wrap everything that should be hidden while it loads. This will avoid calling/render any unneeded method until it loads. Something like:
```jsx
<LoadableBlock isLoading={ isLoading } numLines={ 20 }>
```
And simplify all the get* and format* functions to not check if it's loading.

Reply:
Thanks! I had put a really bad workaround in place for this while I fixed other issues. Will update to use the `LoadableBlock` solution 🙂

Reply:
Updated in 3df4f0d0fc11527e705063fd02d49d051a3e9fd4.

However, this is a little imperfect because the select briefly displays before the LoadableBlock element appears - I'm guessing this is because `isLoading` is false for a brief moment before the resolver starts loading the data. Have you encountered this before?

Reply:
Yep! It returns false before start loading it. We can use `hasFinishedResolution` to mitigate that. Something like:
```js
isLoading: isResolving( 'getBusinessTypes' ) || ! hasFinishedResolution( 'getBusinessTypes' )
```